platform/pulp/platform/models/progress.py

```diff
@@ -0,0 +1,195 @@
+        :type f:  function
+
+        :param args: The positional arguments to be passed to func
+        :type args:  list
+
+        :param kwargs: The keyword arguments to be passed to func
+        :type kwargs:  dict
+
+        :return: The return value of func(*args, **kwargs)
+        """
+        try:
+            if self.state != self.RUNNING:
+                self.state = self.RUNNING
+                self.state.save()
+            return func(*args, **kwargs)
+        except Exception:
+            self.state = self.ERROR
+            self.save()
+            raise
+        finally:
+            self.state = self.FINISHED
+            self.save()
+
+
+class ProgressSpinner(ProgressReportingBase):
+    """
+    Shows progress reporting when the count of items is not known.
+
+    Plugin writers should create these objects to show progress reporting of a single step or
+    aspect of work which has a name and a state. For example:
+
+    ProgressSpinner(message='Publishing Metadata') # default state is 'waiting'
+
+    ProgressSpinner(message='Publishing Metadata', state='running') # specify 'running' instead
+
+    # Update the state to FINISHED and save it
+    metadata_progress = ProgressSpinner(message='Publishing Metadata', state='running')
+    metadata_progress.state = 'finished'
+    metadata_progress.save()
+
+    The call_and_observe() method simplifies state management by providing automatic state
+    transitions for the RUNNING FINISHED and ERROR states.
+
+    ProgressSpinner objects are associated with a Task and auto-discover and populate the task id
+    when saved.
+    """
+    pass
+
+
+class ProgressBar(ProgressReportingBase):
+    """
+    Shows progress reporting when the count of items is known.
+
+    Plugin writers should create these objects to show progress reporting of a single step or
+    aspect of work which has a name, state, max, and current. For example:
+
+    ProgressBar(message='Publishing files to web', max=23) # default: state='waiting' and current=0
+
+    ProgressBar(message='Publishing files to web', max=23, state='running') # specify the state
+    ProgressBar(message='Publishing files to web', max=23, current=16) # already completed 16
+
+    # Update the state to FINISHED and save it
+    web_progress = ProgressBar('Publishing files to web', max=23, state='running')
+    web_progress.state = 'finished'
+    web_progress.save()
+
+    The next() method can be called with each ...
+
+    It also support iter()...
+
+    ProgressBar objects are associated with a Task and auto-discover and populate the task id when
+    saved.
+
+        Fields:
+
+    :cvar max: The total count of items to be handled by the ProgressBar (required)
+    :type max: models.IntegerField
+
+    :cvar current: The count of items already processed. Defaults to 0.
+    :type current: models.IntegerField
+    """
+    max = models.IntegerField()
+    current = models.IntegerField(default=0)
+
+    def next(self):
```

Comment:
s/next/increment/   Call it this because it doesn't implement the Python iterator interface since it never raises `StopIterator`